sys/sema/sema.c

```diff
@@ -90,24 +85,22 @@ int sem_wait_timed_msg(sem_t *sem, timex_t *timeout, msg_t *msg)
         n.priority = (uint32_t)sched_active_thread->priority;
         n.data = (unsigned int)sched_active_pid;
         n.next = NULL;
-        priority_queue_add(&sem->queue, &n);
+        priority_queue_add(&sema->queue, &n);
 
-        DEBUG("sem_wait: %" PRIkernel_pid ": Adding node to semaphore queue: prio: %" PRIu32 "\n",
+        DEBUG("sema_wait: %" PRIkernel_pid ": Adding node to semaphore queue: prio: %" PRIu32 "\n",
               sched_active_thread->pid, sched_active_thread->priority);
 
-        if (timeout != NULL) {
-            vtimer_set_msg(&timeout_timer, *timeout, sched_active_pid,
-                           MSG_TIMEOUT, sem);
+        if (timeout != 0) {
+            timeout_msg.type = MSG_TIMEOUT;
+            timeout_msg.content.ptr = (char *)sema;
+            /* we will stay in the same stack context so we can use timeout_msg */
+            xtimer_set_msg64(&timeout_timer, timeout, &timeout_msg, sched_active_pid);
         }
 
         restoreIRQ(old_state);
         msg_receive(msg);
```

Comment:
it is a good idea to explicitly remove the used xtimer before returning from sema_wait_timed_msg as the message `msg` may have originated elsewhere (regardless of content). Both because the xtimer object has local scope, and also in order to prevent future spurious messages from xtimer.